IOperation Test Porting Pt 3

333fred opened the pull request:
This ports the last of the commits from https://github.com/dotnet/roslyn/pull/49534 and addresses feedback on them. ~~The only remaining test hook failures in master were introduced by the UsedAssemblyReferences branch and Aleksey is taking a look at those. If that isn't fixed by the time this is merged, I'll turn those tests off for the test hook when I enable the test hook in CI.~~ After investigation, these failures are invalid, and the assert is removed by the last commit in this PR.

As in my previous porting PRs, ported commits are prefixed with Port:, and fixup commits are prefixed with Fixup:. There are 4 new, previously unreviewed commits at the end that are not prefixed. See the commit descriptions for what each one does.

333fred commented on src/Compilers/CSharp/Portable/Binder/Binder_Attributes.cs:
```diff
@@ -180,11 +180,20 @@ private BoundAttribute BindAttributeCore(AttributeSyntax node, NamedTypeSymbol a
+
+            // If the arguments didn't have a natural type, they're not constants, and so would already have diagnostics from the rest of binding.
+            var ignoredDiagnostics = new DiagnosticBag();
+            ImmutableArray<BoundExpression> boundConstructorArguments = constructorArguments.Arguments.SelectAsArray(
```
>constructorArguments.Arguments [](start = 72, length = 30)

AlekseyTs said:

> Are we leaking this builder?

333fred replied:
Changing this back to resolved so you see my update here: `AnalyzedArguments.Arguments` are not pooled. There is no need to free that builder.

---
In reply to: [575600283](https://github.com/dotnet/roslyn/pull/51206#discussion_r575600283) [](ancestors = 575600283)

333fred commented on src/Compilers/CSharp/Portable/Binder/Binder_Attributes.cs:
```diff
@@ -180,11 +180,20 @@ private BoundAttribute BindAttributeCore(AttributeSyntax node, NamedTypeSymbol a
+            // If the arguments didn't have a natural type, they're not constants, and so would already have diagnostics from the rest of binding.
+            var ignoredDiagnostics = new DiagnosticBag();
+            ImmutableArray<BoundExpression> boundConstructorArguments = constructorArguments.Arguments.SelectAsArray(
+                (arg, thisAndDiagnostics) => thisAndDiagnostics.@this.BindToNaturalType(arg, thisAndDiagnostics.ignoredDiagnostics),
```
>BindToNaturalType [](start = 70, length = 17)

AlekseyTs said:

> `BindToTypeForErrorRecovery`?
> We also have a `BuildArgumentsForErrorRecovery`, see if we can use that helper.

333fred replied:
Done. In order to use that helper, I had to split up the initialization into 2 versions: one where we could use the helper because we found a constructor, and one where we just bind to the natural type if it exists for the case where we didn't find a constructor.

---
In reply to: [575602126](https://github.com/dotnet/roslyn/pull/51206#discussion_r575602126) [](ancestors = 575602126)

333fred commented on src/Compilers/CSharp/Portable/Binder/Binder_Attributes.cs:
```diff
@@ -180,11 +180,20 @@ private BoundAttribute BindAttributeCore(AttributeSyntax node, NamedTypeSymbol a
+                (@this: this, ignoredDiagnostics));
             ImmutableArray<string> boundConstructorArgumentNamesOpt = constructorArguments.GetNames();
-            ImmutableArray<BoundExpression> boundNamedArguments = analyzedArguments.NamedArguments;
+            ImmutableArray<BoundExpression> boundNamedArguments = analyzedArguments.NamedArguments.SelectAsArray(
```
>analyzedArguments.NamedArguments [](start = 66, length = 32)

AlekseyTs said:

> See if we want to change `BindAttributeArguments` to keep it as a builder since we are going to allocate a new array here.

333fred commented on src/Compilers/CSharp/Portable/Binder/Binder_Attributes.cs:
```diff
@@ -180,11 +180,20 @@ private BoundAttribute BindAttributeCore(AttributeSyntax node, NamedTypeSymbol a
             ImmutableArray<string> boundConstructorArgumentNamesOpt = constructorArguments.GetNames();
-            ImmutableArray<BoundExpression> boundNamedArguments = analyzedArguments.NamedArguments;
+            ImmutableArray<BoundExpression> boundNamedArguments = analyzedArguments.NamedArguments.SelectAsArray(
+                (namedArg, thisAndDiagnostics) => thisAndDiagnostics.@this.BindToNaturalType(namedArg, thisAndDiagnostics.ignoredDiagnostics),
```
>BindToNaturalType [](start = 75, length = 17)

AlekseyTs said:

> `BindToTypeForErrorRecovery`?

333fred commented on src/Compilers/Test/Core/Compilation/ControlFlowGraphVerifier.cs:
```diff
@@ -1110,6 +1110,16 @@ bool isLongLivedCaptureReferenceSyntax(SyntaxNode captureReferenceSyntax)
+
+                                case CSharp.SyntaxKind.SimpleAssignmentExpression:
+                                    if (((CSharp.Syntax.AssignmentExpressionSyntax)syntax.Parent).Left == syntax
+                                        && syntax.Parent.Parent is InitializerExpressionSyntax { Parent: CSharp.Syntax.ObjectCreationExpressionSyntax })
```
>&& syntax.Parent.Parent is InitializerExpressionSyntax { Parent: CSharp.Syntax.ObjectCreationExpressionSyntax }) [](start = 40, length = 112)

AlekseyTs said:

> Would it be possible to tighten the condition by also checking something other than syntax? Perhaps check symbols involved?
> I do not think we should recognize this as long lived, it doesn't escape the statement.
> Could isObjectOrArrayInitializerInitializedObjectTarget help for this scenario?

333fred commented on src/Compilers/Test/Core/Compilation/CompilationExtensions.cs:
```diff
@@ -3,6 +3,8 @@
 
 #nullable disable
+// Uncomment to enable the IOperation test hook on all test runs. Do not commit this uncommented.
+//#define ROSLYN_TEST_IOPERATION
```
This is mainly a helper for running the test hook in VS.

333fred commented:
@dotnet/roslyn-compiler @AlekseyTs for review.

AlekseyTs commented on src/Compilers/CSharp/Portable/Binder/Binder_Attributes.cs:
```diff
@@ -177,12 +179,21 @@ private BoundAttribute BindAttributeCore(AttributeSyntax node, NamedTypeSymbol a
                     Error(diagnostics, ErrorCode.ERR_AttributeCtorInParameter, node, attributeConstructor.ToDisplayString(SymbolDisplayFormat.CSharpErrorMessageFormat));
                 }
+
+                boundConstructorArguments = BuildArgumentsForErrorRecovery(constructorArguments, ImmutableArray.Create(attributeConstructor));
```
>BuildArgumentsForErrorRecovery [](start = 44, length = 30)

It looks like we are calling this non-discriminatively both for error and success cases. This doesn't feel right. It is quite possible I misunderstood what issue we are trying to deal with from the very beginning.

333fred replied:
The case we are dealing with is that, in error cases, there are some arguments that will need to be converted. I can make this conditional on whether there was a lookup failure though.

---
In reply to: [577675907](https://github.com/dotnet/roslyn/pull/51206#discussion_r577675907) [](ancestors = 577675907)

AlekseyTs replied:
> The case we are dealing with is that, in error cases, there are some arguments that will need to be converted. I can make this conditional on whether there was a lookup failure though.

This logic still feels somewhat suspicious to me. Where does conversion happen when lookup succeeds?

---
In reply to: [579374920](https://github.com/dotnet/roslyn/pull/51206#discussion_r579374920) [](ancestors = 579374920,577675907)

333fred replied:
> Where does conversion happen when lookup succeeds?

As part of `BindAttributeConstructor` above. The nested call to `TryPerformConstructorOverloadResolution` coerces the arguments if overload resolution succeeded (and for cases when an inaccessible constructor was found as well).

---
In reply to: [581140844](https://github.com/dotnet/roslyn/pull/51206#discussion_r581140844) [](ancestors = 581140844,579374920,577675907)

AlekseyTs replied:
> As part of BindAttributeConstructor above. The nested call to TryPerformConstructorOverloadResolution coerces the arguments if overload resolution succeeded (and for cases when an inaccessible constructor was found as well).

Then it feels to me that this operation should be performed based on whether ```TryPerformConstructorOverloadResolution``` succeeded and based on ```candidateConstructors``` that it produces, rather than based on ```resultKind```. Relying on ```resultKind``` feels fragile, exact correlation between its value and the failure ```TryPerformConstructorOverloadResolution``` is not obvious. Also, I think that ```SelectAsArrayInPlaceAndFree``` below should be called only when ```BindAttributeConstructor``` was not called.

---
In reply to: [582325554](https://github.com/dotnet/roslyn/pull/51206#discussion_r582325554) [](ancestors = 582325554,581140844,579374920,577675907)

333fred replied:
Sure, that seems like a reasonable change. I'll do that.

---
In reply to: [582350196](https://github.com/dotnet/roslyn/pull/51206#discussion_r582350196) [](ancestors = 582350196,582325554,581140844,579374920,577675907)

AlekseyTs commented on src/Compilers/CSharp/Portable/Binder/Binder_Attributes.cs:
```diff
@@ -177,12 +179,21 @@ private BoundAttribute BindAttributeCore(AttributeSyntax node, NamedTypeSymbol a
             ImmutableArray<string> boundConstructorArgumentNamesOpt = constructorArguments.GetNames();
-            ImmutableArray<BoundExpression> boundNamedArguments = analyzedArguments.NamedArguments;
+            ImmutableArray<BoundExpression> boundNamedArguments =
+                analyzedArguments.NamedArguments?.SelectAsArrayInPlaceAndFree((namedArg, @this) => @this.BindToTypeForErrorRecovery(namedArg), this)
```
>@this.BindToTypeForErrorRecovery(namedArg) [](start = 99, length = 42)

Similar comment here. It doesn't feel appropriate to do this for success cases.

333fred replied:
I've looked into this, and we actually don't need to do this at all. These are always `BoundAssignmentOperator` nodes, and the right sides are always appropriately converted by binding. I've changed the code to strongly assume that these are always `BoundAssignmentOperator`s and simply assert that the `Right` never needs binding here.

---
In reply to: [577678797](https://github.com/dotnet/roslyn/pull/51206#discussion_r577678797) [](ancestors = 577678797)

AlekseyTs commented on src/Compilers/Core/Portable/Operations/ControlFlowGraphBuilder.cs:
```diff
@@ -3310,12 +3315,31 @@ static bool isConditionalAccessInstancePresentInChildren(IOperation operation)
                     {
                         return true;
                     }
+                    else if (operation is InvalidOperation invalidChild)
```
>else if (operation is InvalidOperation invalidChild) [](start = 20, length = 52)

I am not sure why do we need to do this this check inside the loop. Are we checking this at the beginning of the function?

333fred replied:
Typo, that should have been `enumerator.Current`. Adding a test to demonstrate.

---
In reply to: [577691937](https://github.com/dotnet/roslyn/pull/51206#discussion_r577691937) [](ancestors = 577691937)

AlekseyTs commented:
Done with review pass (commit 9)

333fred commented:
@AlekseyTs addressed feedback. I've split the resolutions into separate commits as they followup on different iteration comments.

---
In reply to: [780630534](https://github.com/dotnet/roslyn/pull/51206#issuecomment-780630534) [](ancestors = 780630534)

333fred commented:
@AlekseyTs @dotnet/roslyn-compiler for review.

AlekseyTs commented:
Done with review pass (commit 11)

333fred commented:
@AlekseyTs addressed feedback.

---
In reply to: [784305142](https://github.com/dotnet/roslyn/pull/51206#issuecomment-784305142) [](ancestors = 784305142)

AlekseyTs reviewed (approved):
LGTM (commit 13), assuming CI is passing.

333fred commented:
@dotnet/roslyn-compiler for a second review please.

jcouv commented on src/Compilers/CSharp/Portable/Symbols/TypeWithAnnotations.cs:
```diff
@@ -942,8 +942,6 @@ private TypeSymbol GetResolvedType()
             {
                 if ((object)_resolved == null)
                 {
-                    Debug.Assert(_underlying.IsSafeToResolve());
```
Why remove this assertion?

jcouv commented:
>         /// The result of this method captures some AnalyzedArguments, which must be free'ed by the caller.

Please update the comment (the result now also holds pooled builders which the caller is responsible for freeing)

---
Refers to: src/Compilers/CSharp/Portable/Binder/Binder_Attributes.cs:305 in 76e0ccd. [](commit_id = 76e0ccd279e1efb376d41bcab8414794a30670d1, deletion_comment = False)

jcouv commented on src/Compilers/CSharp/Portable/Binder/Binder_Attributes.cs:
```diff
@@ -165,8 +166,16 @@ private BoundAttribute BindAttributeCore(AttributeSyntax node, NamedTypeSymbol a
+            else
+            {
+                boundConstructorArguments = analyzedArguments.ConstructorArguments.Arguments.SelectAsArray(
+                    (arg, attributeArgumentBinder) => attributeArgumentBinder.BindToTypeForErrorRecovery(arg),
```
>(arg, attributeArgumentBinder) [](start = 20, length = 30)

static?

jcouv reviewed (commented):
Done with review pass (iteration 13)

jcouv reviewed (approved):
LGTM Thanks (iteration 14)
